std/rb.zig

```diff
@@ -0,0 +1,523 @@
+const Color = enum(u1) {
+    Black,
+    Red,
+};
+const Red = Color.Red;
+const Black = Color.Black;
+
+pub const Node = struct {
+    left: ?*Node,
+    right: ?*Node,
+    parent_and_color: usize,
+
+    pub fn next(constnode: *Node) ?*Node {
+        var node = constnode;
+
+        if (node.right) |right| {
+            var n = right;
+            while (n.left != null)
+                n = n.left.?;
+            return n;
+        }
+
+        while (true) {
+            var parent = node.get_parent();
+            if (parent) |p| {
+                if (node != p.right)
+                    return p;
+                node = p;
+            } else
+                return null;
+        }
+    }
+
+    pub fn prev(constnode: *Node) ?*Node {
+        var node = constnode;
+
+        if (node.left) |left| {
+            var n = left;
+            while (n.right != null)
+                n = n.right.?;
+            return n;
+        }
+
+        while (true) {
+            var parent = node.get_parent();
+            if (parent) |p| {
+                if (node != p.left)
+                    return p;
+                node = p;
+            } else
+                return null;
+        }
+    }
+
+    pub fn is_root(node: *Node) bool {
+        return node.get_parent() == null;
+    }
+
+    fn is_red(node: *Node) bool {
+        return node.get_color() == Red;
+    }
+
+    fn is_black(node: *Node) bool {
+        return node.get_color() == Black;
+    }
+
+    fn set_parent(node: *Node, parent: ?*Node) void {
+        node.parent_and_color = @ptrToInt(parent) | (node.parent_and_color & 1);
+    }
+
+    fn get_parent(node: *Node) ?*Node {
+        const mask: usize = 1;
+        return @intToPtr(*Node, node.parent_and_color & ~mask);
+    }
+
+    fn set_color(node: *Node, color: Color) void {
+        const mask: usize = 1;
+        node.parent_and_color = (node.parent_and_color & ~mask) | @enumToInt(color);
+    }
+
+    fn get_color(node: *Node) Color {
+        return @intToEnum(Color, @intCast(u1, node.parent_and_color & 1));
+    }
+
+    fn set_child(node: *Node, child: ?*Node, is_left: bool) void {
+        if (is_left) {
+            node.left = child;
+        } else {
+            node.right = child;
+        }
+    }
+
+    fn get_first(nodeconst: *Node) *Node {
+        var node = nodeconst;
+        while (node.left) |left| {
+            node = left;
+        }
+        return node;
+    }
+
+    fn get_last(node: *Node) *Node {
+        while (node.right) |right| {
+            node = right;
+        }
+        return node;
+    }
+};
+
+pub const Tree = struct {
+    root: ?*Node,
+    compare_fn: fn(*Node, *Node) i32,
```

Comment:
Would this signature work for `compare_fn`?

`lessThan: fn (lhs: *Node, rhs: *Node) bool`

This is what we use over in `std.sort`.

Reply:
bool doesn't work because we have to return a tri-state like C's memcmp()

Reply:
You can achieve the same thing with a lessThan function. If you want an example you can look at my ordered map (which is not red-black but an anderson tree) https://github.com/BarabasGitHub/ZigZag/blob/master/containers/ordered_map.zig

Reply:
No, I have to be able to detect duplicate keys, because a red-black tree can't have duplicate keys (unless you use it as a priority queue).

Reply:
@shawnl I'd recommend allowing duplicate keys at the user's risk. I believe this is how C++ implements `std::set`. Essentially, `operator<` is the only operator used to compare, and equality is ignored.

Reply:
It doesn't allow duplicate keys, that's what you have `std::multiset` for.

It's quite easy `a == b` is the same as `!(a < b) and !(b < a)` or `!(a < b or b < a)` (or should be,  otherwise your `<` operator is broken)

Reply:
duplicate keys requires alot more code and breaks the O(nlogn) behavior. (unless it is a priority queue, in case the reading behavior is limited and well-defined).

Reply:
On the other hand, a rb-tree is superior to a B-tree (which does allow duplicate keys) because it doesn't require any dynamic memory allocation.

Reply:
Sorry I don't mean the data structure allows for it. And yes @BarabasGitHub my mistake, that is what C++ does. @shawnl whatever the case, I'd love to see a common operator that could be used consistently throughout the Zig standard library so we don't need both functions.

Reply:
Ya C++20 is adding that too. If we go that route, we should change `std.sort` too.

Reply:
If it's not going to be a `lessThan` function, then why would it be `isize` and not `enum {Equal, LessThan, GreaterThan}`?

Reply:
where should the definition of this enum reside?

Reply:
Doesn't spaceship operator only make sense when you have operator overloading?

Reply:
Spaceship operator was just an analogy. You would implement it with a comparator function here that returns that enum type.